ffi/src/lib.rs

```diff
@@ -0,0 +1,180 @@
+use anyhow::bail;
+use anyhow::Result;
+use memcache::Client;
+use memcache::CommandError::KeyExists;
+use memcache::MemcacheError::CommandError;
+use memcache::Url;
+use std::cell::Cell;
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::ffi::CStr;
+use std::ffi::CString;
+use std::str;
+use std::time::SystemTime;
+
+struct Rl {
+    url: Url,
+    client: Option<Client>,
+    error: CString,
+}
+
+impl Rl {
+    fn new(url: &CStr) -> Result<Rl> {
+        let url = Url::parse(url.to_str()?)?;
+        Ok(Rl {
+            url: url,
+            client: None,
+            error: CString::new("")?,
+        })
+    }
+
+    fn rate_limit(
+        &mut self,
+        prefix: &CStr,
+        size: u32,
+        rate_max: u32,
+        rate_seconds: u32,
+    ) -> Result<bool> {
+        let prefix = prefix.to_str()?;
+
+        /* re-connect if necessary */
+        if let None = self.client {
+            self.client = Some(Client::connect(String::from(self.url.clone()))?)
+        };
+
+        if let Some(client) = &self.client {
+            let now = SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)?
+                .as_secs();
+
+            let key = format!("{}{}", prefix, now);
+            match client.add(&key, 0, rate_seconds + 1) {
+                Ok(()) => (),
+                Err(e) => match e {
+                    CommandError(e) if e == KeyExists => (),
+                    e => bail!(e),
+                },
+            };
+
+            let keys: Vec<String> = (0..rate_seconds)
+                .map(|n| format!("{}{}", prefix, n))
+                .collect();
+            let keys: Vec<&str> = keys.iter().map(|s| &**s).collect();
```

Comment:
Seems like I had to iterate over this twice to get a `Vec<String>`.  Trying to convert it inside the map gave a temporary variable error.  But it seems like there should be a better way to do this?